src/v/cluster/archival/segment_reupload.cc

```diff
@@ -273,6 +276,42 @@ void segment_collector::align_end_offset_to_manifest(
     }
 }
 
+void segment_collector::align_with_term_boundary() {
+    auto opt_term = _log.get_term(_begin_inclusive);
+    if (!opt_term.has_value()) {
+        vlog(
+          archival_log.warn,
+          "Can't find term for offset {}, the log could be truncated",
+          _begin_inclusive);
+        throw std::runtime_error(fmt_with_ctx(
+          fmt::format, "Can't find term for offset {}", _begin_inclusive));
+    }
+    auto opt_max_offset = _log.get_term_last_offset(opt_term.value());
+    if (!opt_max_offset.has_value()) {
+        vlog(
+          archival_log.warn,
+          "Can't find last offset for term {}",
+          opt_term.value());
+        throw std::runtime_error(fmt_with_ctx(
+          fmt::format, "Can't find last offset for term {}", opt_term.value()));
+    }
+    vlog(
+      archival_log.debug,
+      "align_with_term_boundary, begin={}, end={}, term={}, "
+      "last-term-offset={}",
+      _begin_inclusive,
+      _end_inclusive,
+      opt_term.value(),
+      opt_max_offset.value());
+    // Invariant: 'max_offset' can't be in the middle of the segment
```

Comment:
This comment is ambiguous. Why is this variant is important and when? `max_offset` can be in the "middle" of a segment if we're scanning through the "current" active segment, no?